### PR TITLE
Upgrade Gradle wrapper version and use distribution that includes sources instead of binary only #7436

### LINKED ISSUE #7436: Upgrade Gradle wrapper version and use distribution that includes sources instead of binary only
1. Gradle has been making several improvements in terms of usability and performance since version `2.12`. We should upgrade to version `3.5`.

   Some examples of improvements include:
[Tooling API generates more progress events](https://docs.gradle.org/3.3/release-notes.html#tooling-api-generates-more-progress-events)
[Production ready incremental builds](https://docs.gradle.org/3.4/release-notes.html#stable-java-incremental-compilation)
[More details in console output](https://docs.gradle.org/current/release-notes.html#console-upgrade)

2. Use the `all` distribution wrapper to provide IDE with Gradle API/DSL documentation.

### PR DESCRIPTION
**Pull request by LiHaoTan**
Fixes #7436

**Outline of Solution**
Upgrade Gradle wrapper version to make use of improvements introduced.

Some examples of improvements include:
[Tooling API generates more progress events](https://docs.gradle.org/3.3/release-notes.html#tooling-api-generates-more-progress-events)
[Production ready incremental builds](https://docs.gradle.org/3.4/release-notes.html#stable-java-incremental-compilation)
[More details in console output](https://docs.gradle.org/current/release-notes.html#console-upgrade)

### CODE REVIEW

**Comment by LiHaoTan** [comment]
@wkurniawan07 I did not test it out in Eclipse but we should at least be able to get access to quick documentation in IDEs easily.

In my opinion, it is also useful to access sources easily given how much "magic" Gradle adds but that is arguable.

Anyway, the increase in size is negligible as well `53,639` bytes to `54,783` bytes.

File: build.gradle [hunk 1]
```diff
@@ -431,8 +443,9 @@ def afterTestClosure = { descriptor, result ->
     }
 }
 
+import org.gradle.internal.serialize.PlaceholderException
 def checkConfigFailureClosure = { descriptor, result ->
-    if (result.exception instanceof org.gradle.messaging.remote.internal.PlaceholderException
+    if (result.exception instanceof PlaceholderException
```
Comment (wkurniawan07) [inline build.gradle]: Do one of the following:
- Empty line between `import ...` and `def ...`
- Use fully qualified name as like before, only changing the package name as necessary

**Comment by wkurniawan07** [comment]
Anyway, @whipermr5 have a review as well.

File: build.gradle (hunk 1 above)
Comment (LiHaoTan) [inline build.gradle]: changed

File: build.gradle [hunk 2]
```diff
@@ -187,49 +187,55 @@ task createConfigs {
...
+                ".templates/intellijRunConfigs/Local_Tests.xml"
+        ]
+        templatesToCopy.findAll {
+            !(new File(it.replace(".templates/intellijRunConfigs", ".idea/runConfigurations"))).exists()
+        }.each {
+            filename ->
+                copy {
+                    from filename
+                    into ".idea/runConfigurations"
+                }
```
Comment (whipermr5) [inline build.gradle]:
```
}.each { filename ->
    copy {
        from filename
        into ".idea/runConfigurations"
    }
}
```

File: build.gradle [hunk 3]
```diff
@@ -187,49 +187,55 @@ task createConfigs {
...
+            filename ->
+                copy {
+                    from filename
+                    into ".launches"
+                    rename '(.*).xml', '$1'
+                }
+                filename = filename.replace(".templates/eclipseLaunches", ".launches").replace(".xml", "")
+                def fileContents = new File(filename).getText("UTF-8")
+                def projectName = "${projectDir}".replaceAll(/.*(\/|\\)/, "")
+                new File(filename).text = new File(filename).getText("UTF-8").replaceAll(/\$\{project\.name\}/, projectName)
```
Comment (whipermr5) [inline build.gradle]:
```
}.each { filename ->
    copy {
        from filename
        into ".launches"
        rename '(.*).xml', '$1'
    }
    filename = filename.replace(".templates/eclipseLaunches", ".launches").replace(".xml", "")
    def fileContents = new File(filename).getText("UTF-8")
    def projectName = "${projectDir}".replaceAll(/.*(\/|\\)/, "")
    new File(filename).text = new File(filename).getText("UTF-8").replaceAll(/\$\{project\.name\}/, projectName)
}
```

**Comment by LiHaoTan** [comment]
@whipermr5 done

**Pull request merged**